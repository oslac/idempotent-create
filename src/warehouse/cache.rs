use crate::error::get_error_cause;
use crate::ikey::IKey;
use crate::user::User;

use axum::http::StatusCode;
use color_eyre::Report;
use std::collections::HashMap;
use std::fmt::Debug;
use std::fmt::Display;

/// A response cache, mapping client provided [IKey] to [CachedResponse].
#[derive(Clone, Debug, Default)]
pub struct Cache(HashMap<IKey, CachedResponse>);

#[derive(Clone, Debug)]
pub struct CachedResponse {
    pub status: StatusCode,
    pub user: User,
}

#[derive(thiserror::Error)]
pub enum CacheError {
    #[error("Cache for {0}")]
    CacheMiss(String),
    #[error(transparent)]
    Unexpected(#[from] Report),
}

impl Cache {
    pub fn new() -> Self {
        Self(HashMap::default())
    }

    /// Given a [IKey] and [CachedResponse], performs an upsert.
    pub async fn set(&mut self, key: &IKey, res: &CachedResponse) -> Result<(), CacheError> {
        let res = res.clone();
        let key = key.clone();
        self.0.insert(key, res);
        Ok(())
    }

    /// Given an [IKey], either returns a [CachedResponse] on a cache hit, or
    /// [CacheError] on miss.
    pub async fn get(&self, key: &IKey) -> Result<CachedResponse, CacheError> {
        let res = self.0.get(key).ok_or(CacheError::CacheMiss(key.to_string()))?;
        Ok(res.clone())
    }
}

impl Debug for CacheError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        get_error_cause(self, f)
    }
}

impl Display for CachedResponse {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Cached ({}, {})", self.status, self.user.id)
    }
}
