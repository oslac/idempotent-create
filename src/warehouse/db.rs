use crate::error::get_error_cause;
use crate::user::NewUser;
use crate::user::User;

use color_eyre::Report;
use std::collections::HashMap;
use std::fmt::Debug;

type UserId = u64;

#[derive(Clone, Debug, Default)]
pub struct UserRepository(HashMap<UserId, User>);

impl UserRepository {
    pub fn new() -> Self {
        Self(HashMap::default())
    }

    /// Create a new [User] from [NewUser].
    pub fn create(&mut self, new_user: &NewUser) -> Result<User, UserRepoError> {
        self.email_is_available(&new_user.email)?;
        tracing::info!("Email Is Free");
        let new_id = self.random_id();
        let new_user = User::new(new_id, new_user.email.to_owned());
        // This will always return ``None``:
        self.0.insert(new_user.id, new_user.clone());
        tracing::info!("New User Inserted to DB");
        Ok(new_user)
    }

    /// Get all users.
    pub fn list(&self) -> Result<Vec<User>, UserRepoError> {
        let users = self.0.values().cloned().collect();
        Ok(users)
    }

    /// Returns [User] with id ``id``; *otherwise* `NotFound`.
    pub fn get(&self, user_id: u64) -> Result<User, UserRepoError> {
        self.0.get(&user_id).cloned().ok_or(UserRepoError::UserNotFound(user_id))
    }

    fn random_id(&self) -> UserId {
        self.0.keys().max().map_or(1, |id| id + 1)
    }

    fn email_is_available(&self, new_email: &str) -> Result<(), UserRepoError> {
        match self.0.values().any(|user| user.email == new_email) {
            true => Err(UserRepoError::EmailTaken(new_email.to_string())),
            false => Ok(()),
        }
    }
}

#[derive(thiserror::Error)]
pub enum UserRepoError {
    #[error("User {0} Not Found")]
    UserNotFound(u64),
    #[error("Email {0} Is Taken")]
    EmailTaken(String),
    #[error(transparent)]
    Internal(#[from] Report),
}

impl Debug for UserRepoError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        get_error_cause(self, f)
    }
}
