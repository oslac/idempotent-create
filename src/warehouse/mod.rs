//! Types related to storage, repositories, caches etc.
mod cache;
mod db;

pub use db::UserRepoError;
pub use db::UserRepository;

pub use cache::Cache;
pub use cache::CacheError;
pub use cache::CachedResponse;
