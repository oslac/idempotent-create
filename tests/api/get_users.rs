use crate::test_app::TestApp;
use lib::warehouse::UserRepository;

use hyper::body::to_bytes as BodyToBytes;
use hyper::StatusCode;
use serde_json::Value;
use tower::ServiceExt;

#[tokio::test]
async fn without_any_users() {
    // I. Arrange
    let app = TestApp::new(UserRepository::new()).await;
    let req = app.get_users();

    // II. Act
    let response = app.router().oneshot(req).await.unwrap();

    // III. Assert
    let expected_status = StatusCode::OK;
    let actual_status = response.status();
    assert_eq!(expected_status, actual_status);

    let expected_content_type = "application/json";
    let actual_content_type = response.headers().get("Content-Type").unwrap();
    assert_eq!(expected_content_type, actual_content_type);

    let actual_body = BodyToBytes(response.into_body()).await.unwrap();
    let actual_body: Value = serde_json::from_slice(&actual_body).unwrap();
    insta::assert_json_snapshot!(&actual_body);
}

#[tokio::test]
async fn with_some_users() {
    // 1. Arrange
    let data = TestApp::init_repo_data();
    let app = TestApp::new(data).await;
    let req = app.get_users();

    // II. Act
    let response = app.router().oneshot(req).await.unwrap();

    // III. Assert
    let expected_status = StatusCode::OK;
    let actual_status = response.status();
    assert_eq!(expected_status, actual_status);

    let expected_content_type = "application/json";
    let actual_content_type = response.headers().get("Content-Type").unwrap();
    assert_eq!(expected_content_type, actual_content_type);

    let actual_body = BodyToBytes(response.into_body()).await.unwrap();
    let actual_body: Value = serde_json::from_slice(&actual_body).unwrap();
    insta::assert_json_snapshot!(&actual_body);
}
