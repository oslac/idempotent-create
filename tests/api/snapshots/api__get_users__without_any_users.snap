---
source: tests/api/get_users.rs
expression: "&actual_body"
---
[]
